File: src/main/java/org/minefortress/entity/ai/controls/PlaceControl.java
package org.minefortress.entity.ai.controls;

import net.minecraft.block.BlockState;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.event.GameEvent;
import net.remmintan.mods.minefortress.core.utils.ServerExtensionsKt;
import net.remmintan.mods.minefortress.core.utils.ServerModUtils;
import org.minefortress.entity.Colonist;
import org.minefortress.tasks.block.info.BlockStateTaskBlockInfo;
import org.minefortress.tasks.block.info.ItemTaskBlockInfo;

import static net.remmintan.mods.minefortress.core.utils.SimilarItemsHelper.isIgnorable;
import static org.minefortress.entity.colonist.FortressHungerManager.ACTIVE_EXHAUSTION;


public class PlaceControl extends PositionedActionControl {

    private final Colonist colonist;

    private float placeCooldown = 0;
    private int failedInteractions = 0;
    private boolean cantPlaceUnderMyself = false;

    public PlaceControl(Colonist colonist) {
        this.colonist = colonist;
    }

    @Override
    public void tick() {
        if(isDone()) return;
        if(!super.canReachTheGoal(colonist) || !colonist.getNavigation().isIdle()) return;

        if(placeCooldown>0) placeCooldown-=1f/colonist.getHungerMultiplier();

        final BlockPos blockPos = colonist.getBlockPos();
        if(blockPos.equals(goal))
            if(!colonist.isWallAboveTheHead())
                colonist.getJumpControl().setActive();
            else
                cantPlaceUnderMyself = true;
        else
            placeBlock();
    }

    @Override
    public void reset() {
        super.reset();
        failedInteractions = 0;
        cantPlaceUnderMyself = false;
    }

    protected void placeBlock() {
        colonist.lookAtGoal();
        colonist.putItemInHand(item);

        if (placeCooldown <= 0) {
            this.colonist.swingHand(Hand.MAIN_HAND);
            colonist.addHunger(ACTIVE_EXHAUSTION);

            if(taskBlockInfo instanceof ItemTaskBlockInfo)
                place((ItemTaskBlockInfo) taskBlockInfo);
            if(taskBlockInfo instanceof BlockStateTaskBlockInfo)
                place((BlockStateTaskBlockInfo) taskBlockInfo);
        }
    }

    private void place(ItemTaskBlockInfo blockInfo) {
        final ItemUsageContext context = blockInfo.getContext();
        final ActionResult interactionResult = item.useOnBlock(context);

        if(interactionResult == ActionResult.CONSUME || failedInteractions > 15) {
            if(interactionResult == ActionResult.CONSUME)
                decreaseResourcesAndAddSpecialBlocksAmount();
            this.reset();
            this.placeCooldown = 6;
        } else {
            failedInteractions++;
        }
    }

    private void place(BlockStateTaskBlockInfo blockInfo) {
        final BlockState stateForPlacement = blockInfo.getState();

        colonist.getWorld().setBlockState(goal, stateForPlacement, 3);
        colonist.getWorld().emitGameEvent(colonist, GameEvent.BLOCK_PLACE, goal);

        decreaseResourcesAndAddSpecialBlocksAmount();

        this.reset();
        this.placeCooldown = 6;
    }

    private void decreaseResourcesAndAddSpecialBlocksAmount() {
        final var taskControl = colonist.getTaskControl();
        if (ServerExtensionsKt.isCreativeFortress(colonist.getServer())) {
            taskControl
                    .getTaskId()
                    .ifPresent(it -> {
                        ServerModUtils.getManagersProvider(colonist).ifPresent(mp -> {
                            final var resourceManager = mp.getResourceManager();
                            if (isIgnorable(item)) {
                                resourceManager.removeItemIfExists(it, item);
                            } else {
                                resourceManager.removeReservedItem(it, item);
                            }
                        });
                    });
        }
    }

    public boolean isCantPlaceUnderMyself() {
        return cantPlaceUnderMyself;
    }
}

File: src/main/java/org/minefortress/entity/ai/controls/PlaceControl.kt
package org.minefortress.entity.ai.controls

import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.world.event.GameEvent
import net.remmintan.mods.minefortress.core.utils.BuildingHelper
import net.remmintan.mods.minefortress.core.utils.SimilarItemsHelper
import net.remmintan.mods.minefortress.core.utils.getManagersProvider
import net.remmintan.mods.minefortress.core.utils.isCreativeFortress
import org.minefortress.entity.Colonist
import org.minefortress.entity.colonist.IFortressHungerManager
import org.minefortress.tasks.block.info.BlockStateTaskBlockInfo
import org.minefortress.tasks.block.info.ItemTaskBlockInfo
import org.minefortress.tasks.block.info.ReplaceTaskBlockInfo

class PlaceControl(private val colonist: Colonist) : PositionedActionControl() {
    private var placeCooldown = 0f
    private var failedInteractions = 0
    var isCantPlaceUnderMyself: Boolean = false
        private set

    override fun tick() {
        if (isDone) return
        if (!super.canReachTheGoal(colonist) || !colonist.navigation.isIdle) return

        if (placeCooldown > 0) placeCooldown -= 1f / colonist.hungerMultiplier

        if (colonist.blockPos == goal) {
            if (!colonist.isWallAboveTheHead)
                colonist.jumpControl.setActive()
            else
                isCantPlaceUnderMyself = true
        } else {
            if (taskBlockInfo is ReplaceTaskBlockInfo && !BuildingHelper.canPlaceBlock(colonist.world, goal))
                return // waiting until we remove the block
            placeBlock()
        }
    }

    override fun reset() {
        super.reset()
        failedInteractions = 0
        isCantPlaceUnderMyself = false
    }

    private fun placeBlock() {
        colonist.lookAtGoal()
        colonist.putItemInHand(item)

        if (placeCooldown <= 0) {
            colonist.swingHand(Hand.MAIN_HAND)
            colonist.addHunger(IFortressHungerManager.ACTIVE_EXHAUSTION)

            val blockInfo = taskBlockInfo
            if (blockInfo is ItemTaskBlockInfo) place(blockInfo)
            if (blockInfo is BlockStateTaskBlockInfo) place(blockInfo)
            if (blockInfo is ReplaceTaskBlockInfo) place(blockInfo)
        }
    }

    private fun place(blockInfo: ItemTaskBlockInfo) {
        val context = blockInfo.context
        val interactionResult = item.useOnBlock(context)

        if (interactionResult == ActionResult.CONSUME || failedInteractions > 15) {
            if (interactionResult == ActionResult.CONSUME) decreaseResourcesAndAddSpecialBlocksAmount()
            this.reset()
            this.placeCooldown = 6f
        } else {
            failedInteractions++
        }
    }

    private fun place(blockInfo: BlockStateTaskBlockInfo) {
        val stateForPlacement = blockInfo.state

        colonist.world.setBlockState(goal, stateForPlacement, 3)
        colonist.world.emitGameEvent(colonist, GameEvent.BLOCK_PLACE, goal)

        decreaseResourcesAndAddSpecialBlocksAmount()

        this.reset()
        this.placeCooldown = 6f
    }

    private fun place(blockInfo: ReplaceTaskBlockInfo) {
        val stateForPlacement = blockInfo.state

        colonist.world.setBlockState(goal, stateForPlacement, 3)
        colonist.world.emitGameEvent(colonist, GameEvent.BLOCK_PLACE, goal)

        decreaseResourcesAndAddSpecialBlocksAmount()

        this.reset()
        this.placeCooldown = 6f
    }

    private fun decreaseResourcesAndAddSpecialBlocksAmount() {
        val taskControl = colonist.taskControl
        if (colonist.server.isCreativeFortress()) {
            val taskPos = taskControl.currentTaskPos ?: return
            val fortressPos = colonist.fortressPos ?: return
            val rh = colonist.server.getManagersProvider(fortressPos).resourceHelper
            rh.payItemFromTask(taskPos, item, SimilarItemsHelper.isIgnorable(item))
        }
    }
}
